File: py/glyph/canon.py
"""Canonical JSON profile ``glyph-canon-json-1.1.0`` (see ``SPEC-CANON.md``).

The only byte form GLYPH hashes. ``fingerprint``, the patch ``base`` and the GS1
state hash are all ``sha256(canon_json(v))``. GLYPH text is a renderer and is
never hashed.
"""

from __future__ import annotations

import base64
import hashlib
import json
import math
from typing import List, Sequence

from .loose import (
    MAX_SAFE_INT,
    TENSOR_DTYPE_BITS,
    canon_float,
    canon_time,
    from_json_loose,
)
from .types import GType, GValue, MapEntry

MAX_DEPTH = 1000


class CanonError(ValueError):
    """Value cannot be canonicalized: int outside ±(2^53-1), non-finite float,
    duplicate object key, or nesting deeper than MAX_DEPTH."""


def canon_json(v: GValue) -> str:
    """Canonical JSON text of *v* (SPEC-CANON.md §1-§3)."""
    return _canon(v, 0)


def fingerprint(v: GValue) -> str:
    """The one digest: 64 lowercase hex of sha256(canon_json(v)) (SPEC-CANON.md §5)."""
    return hashlib.sha256(canon_json(v).encode("utf-8")).hexdigest()


def is_canonical(b: bytes) -> bool:
    """True iff *b* is exactly the canonical JSON of the value it encodes.

    Receivers at trust boundaries (patch ingest, GS1 state frames) reject bytes
    that fail this: "exactly one valid encoding" with a stdlib parser.
    """
    if isinstance(b, str):
        b = b.encode("utf-8")
    try:
        return (
            canon_json(from_json_loose(json.loads(b.decode("utf-8")))).encode("utf-8")
            == b
        )
    except (ValueError, OverflowError, RecursionError):
        # JSONDecodeError, UnicodeDecodeError, CanonError, bridge limits,
        # float() overflow on huge ints, and RecursionError on nesting past
        # the interpreter/bridge depth guards — all mean "not canonical".
        return False


def tensor_ref(dtype: str, shape: Sequence[int], data: bytes) -> GValue:
    """``{"$tensor":{dtype,shape,sha256}}`` for raw element bytes (SPEC-CANON.md §4).

    Only sha256(data) enters the value. Raises ValueError for an unknown dtype,
    a shape that is not non-negative ints, or data that is not the packed size
    dtype and shape imply (little-endian, row-major, sub-byte dtypes LSB-first).
    """
    bits = TENSOR_DTYPE_BITS.get(dtype)
    if bits is None:
        raise ValueError(f"unknown tensor dtype {dtype!r}")
    n = 1
    for d in shape:
        # Shape dims are strict ints: floats (even integral ones) and bools
        # are rejected, matching the $tensor bridge (_is_dim in loose.py).
        if isinstance(d, bool) or not isinstance(d, int):
            raise ValueError(f"tensor shape dim must be an int, got {d!r}")
        if d < 0:
            raise ValueError(f"tensor shape has negative dim {d}")
        n *= d
    want = (n * bits + 7) // 8
    if len(data) != want:
        raise ValueError(
            f"tensor data is {len(data)} bytes; dtype {dtype} shape {list(shape)} packs to {want}"
        )
    # Sub-byte dtypes pack LSB-first, so the unused high bits of the last
    # byte are padding and must be zero — otherwise two byte strings would
    # name different tensors with the same element sequence.
    if data and (n * bits) % 8:
        if data[-1] >> ((n * bits) % 8):
            raise ValueError(
                f"tensor data has non-zero padding bits for dtype {dtype} shape {list(shape)}"
            )
    return from_json_loose(
        {"$tensor": {"dtype": dtype, "shape": list(shape), "sha256": hashlib.sha256(data).hexdigest()}}
    )


def _quote(s: str) -> str:
    # RFC 8785 §3.2.2.2 escaping == json.dumps(ensure_ascii=False).
    return json.dumps(s, ensure_ascii=False)


def _canon(v: GValue, depth: int) -> str:
    t = v.type
    if t == GType.NULL:
        return "null"
    if t == GType.BOOL:
        return "true" if v.as_bool() else "false"
    if t == GType.INT:
        n = v.as_int()
        if not -MAX_SAFE_INT <= n <= MAX_SAFE_INT:
            raise CanonError(f"integer outside ±(2^53-1): {n}")
        return str(n)
    if t == GType.FLOAT:
        f = v.as_float()
        if not math.isfinite(f):
            raise CanonError("non-finite float")
        if f.is_integer() and abs(f) <= MAX_SAFE_INT:
            return str(int(f))
        return canon_float(f)
    if t == GType.STR:
        return _quote(v.as_str())
    if t == GType.BYTES:
        return (
            '{"$bytes":' + _quote(base64.b64encode(v.as_bytes()).decode("ascii")) + "}"
        )
    if t == GType.TIME:
        return '{"$time":"' + canon_time(v.as_time()) + '"}'
    if t == GType.ID:
        r = v.as_id()
        return '{"$id":[' + _quote(r.prefix) + "," + _quote(r.value) + "]}"
    if t == GType.LIST:
        _check_depth(depth)
        return "[" + ",".join(_canon(x, depth + 1) for x in v.as_list()) + "]"
    if t == GType.MAP:
        return _object(v.as_map(), depth)
    if t == GType.STRUCT:
        return _object(v.as_struct().fields, depth)
    if t == GType.SUM:
        _check_depth(depth)
        s = v.as_sum()
        inner = "null" if s.value is None else _canon(s.value, depth + 1)
        return "{" + _quote(s.tag) + ":" + inner + "}"
    raise CanonError(f"unsupported type {t}")


def _object(entries: List[MapEntry], depth: int) -> str:
    _check_depth(depth)
    # str order == code point order == UTF-8 byte order.
    entries = sorted(entries, key=lambda e: e.key)
    for a, b in zip(entries, entries[1:]):
        if a.key == b.key:
            raise CanonError(f"duplicate key {a.key!r}")
    return (
        "{"
        + ",".join(_quote(e.key) + ":" + _canon(e.value, depth + 1) for e in entries)
        + "}"
    )


def _check_depth(depth: int) -> None:
    if depth >= MAX_DEPTH:
        raise CanonError(f"nesting depth exceeds {MAX_DEPTH}")

File: js/src/canon.ts
/**
 * Canonical JSON profile glyph-canon-json-1.1.0 (see SPEC-CANON.md).
 *
 * The only byte form GLYPH hashes. fingerprint, the patch base and the GS1
 * state hash are all sha256(canonJson(v)). GLYPH text is a renderer and is
 * never hashed.
 */

import { GValue, MapEntry } from './types';
import {
  canonFloat,
  canonTime,
  bytesToBase64,
  fromJsonLoose,
  TENSOR_DTYPE_BITS,
  tensorRefValue,
  checkTensorPadding,
} from './loose';

export const CANON_MAX_DEPTH = 1000;

/**
 * Value cannot be canonicalized: int outside ±(2^53-1), non-finite float,
 * duplicate object key, or nesting deeper than CANON_MAX_DEPTH.
 */
export class CanonError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'CanonError';
  }
}

/** Canonical JSON text of v (SPEC-CANON.md §1-§3). */
export function canonJson(v: GValue): string {
  return canon(v, 0);
}

/**
 * The one digest: 64 lowercase hex of sha256(canonJson(v)) (SPEC-CANON.md §5).
 * Node-only (uses the crypto module); in a browser hash canonJson(v) with
 * crypto.subtle.
 */
export function fingerprint(v: GValue): string {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const { createHash } = require('crypto');
  return createHash('sha256').update(canonJson(v), 'utf8').digest('hex');
}

/**
 * {"$tensor":{dtype,shape,sha256}} for raw element bytes (SPEC-CANON.md §4).
 * Only sha256(data) enters the value. Throws for an unknown dtype, a negative
 * dim, data that is not the packed size dtype and shape imply
 * (little-endian, row-major, sub-byte dtypes LSB-first), or non-zero padding
 * bits in the unused high bits of the last byte. Node-only (crypto).
 */
export function tensorRef(dtype: string, shape: number[], data: Uint8Array): GValue {
  if (!Object.prototype.hasOwnProperty.call(TENSOR_DTYPE_BITS, dtype)) {
    throw new CanonError(`unknown tensor dtype ${JSON.stringify(dtype)}`);
  }
  if (!shape.every((d) => Number.isInteger(d) && d >= 0)) {
    throw new CanonError(`tensor shape must be non-negative ints: ${JSON.stringify(shape)}`);
  }
  const n = shape.reduce((a, d) => a * d, 1);
  const want = Math.ceil((n * TENSOR_DTYPE_BITS[dtype]) / 8);
  if (data.length !== want) {
    throw new CanonError(`tensor data is ${data.length} bytes; dtype ${dtype} shape [${shape}] packs to ${want}`);
  }
  checkTensorPadding(dtype, shape, data);
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const { createHash } = require('crypto');
  return tensorRefValue(dtype, shape, createHash('sha256').update(data).digest('hex'));
}

/**
 * True iff b is exactly the canonical JSON of the value it encodes.
 * Receivers at trust boundaries (patch ingest, GS1 state frames) reject bytes
 * that fail this: "exactly one valid encoding" with a stdlib parser.
 */
export function isCanonical(b: Uint8Array | string): boolean {
  try {
    const text = typeof b === 'string' ? b : new TextDecoder('utf-8', { fatal: true }).decode(b);
    return canonJson(fromJsonLoose(JSON.parse(text))) === text;
  } catch {
    return false;
  }
}

function quote(s: string): string {
  // RFC 8785 §3.2.2.2 escaping == JSON.stringify.
  return JSON.stringify(s);
}

function canon(v: GValue, depth: number): string {
  switch (v.type) {
    case 'null':
      return 'null';
    case 'bool':
      return v.asBool() ? 'true' : 'false';
    case 'int': {
      const n = v.asInt();
      if (!Number.isSafeInteger(n)) throw new CanonError(`integer outside ±(2^53-1): ${n}`);
      return String(n);
    }
    case 'float': {
      const f = v.asFloat();
      if (!Number.isFinite(f)) throw new CanonError('non-finite float');
      if (Number.isSafeInteger(f)) return String(f); // 1.0 -> "1", -0 -> "0"
      return canonFloat(f);
    }
    case 'str':
      return quote(v.asStr());
    case 'bytes':
      return `{"$bytes":${quote(bytesToBase64(v.asBytes()))}}`;
    case 'time':
      return `{"$time":"${canonTime(v.asTime())}"}`;
    case 'id': {
      const r = v.asId();
      return `{"$id":[${quote(r.prefix)},${quote(r.value)}]}`;
    }
    case 'list':
      checkDepth(depth);
      return '[' + v.asList().map((x) => canon(x, depth + 1)).join(',') + ']';
    case 'map':
      return object(v.asMap(), depth);
    case 'struct':
      return object(v.asStruct().fields, depth);
    case 'sum': {
      checkDepth(depth);
      const s = v.asSum();
      return `{${quote(s.tag)}:${s.value === null ? 'null' : canon(s.value, depth + 1)}}`;
    }
  }
  throw new CanonError(`unsupported type ${(v as GValue).type}`);
}

function object(entries: MapEntry[], depth: number): string {
  checkDepth(depth);
  const sorted = entries.slice().sort((a, b) => compareCodePoints(a.key, b.key));
  const parts: string[] = [];
  for (let i = 0; i < sorted.length; i++) {
    if (i > 0 && sorted[i].key === sorted[i - 1].key) {
      throw new CanonError(`duplicate key ${quote(sorted[i].key)}`);
    }
    parts.push(quote(sorted[i].key) + ':' + canon(sorted[i].value, depth + 1));
  }
  return '{' + parts.join(',') + '}';
}

/**
 * Code point order (== UTF-8 byte order, what Go and Python sort by). Plain
 * string comparison would sort by UTF-16 unit, putting non-BMP keys
 * (surrogates D800-DFFF) below E000-FFFF.
 */
export function compareCodePoints(a: string, b: string): number {
  const n = Math.min(a.length, b.length);
  for (let i = 0; i < n; i++) {
    let x = a.charCodeAt(i);
    let y = b.charCodeAt(i);
    if (x !== y) {
      if (x >= 0xd800 && x <= 0xdfff) x += 0x2800;
      if (y >= 0xd800 && y <= 0xdfff) y += 0x2800;
      return x - y;
    }
  }
  return a.length - b.length;
}

function checkDepth(depth: number): void {
  if (depth >= CANON_MAX_DEPTH) throw new CanonError(`nesting depth exceeds ${CANON_MAX_DEPTH}`);
}
